src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -469,6 +581,74 @@ private static IEnumerable<string> GetPossibleMatchesForSelectedOperation(IEdmOp
             yield return qualifiedContainerName + operationNameWithParameters;
         }
 
+        private bool IsValidExpandToken(string item)
+        {
+            // Check #1: Expand token must contain '(' and end with ')', must be related to navigation property.
+            int idxLP = item.IndexOf('(');
+            if (idxLP == -1
+                || !item.EndsWith(")", StringComparison.Ordinal)
+                || (this.structuredType != null && IsBoundOperationToken(item.Substring(0, idxLP))))
+            {
+                // selected item is not properly parenthesized, or is an operation token.
+                return false;
+            }
+
+            // Check #2: parentheses must be balanced.
+            int count = 0;
+            foreach (char c in item)
+            {
+                if (c == '(')
+                {
+                    ++count;
+                }
+                else if (c == ')')
+                {
+                    --count;
+                }
+
+                if (count < 0)
+                {
+                    // Malformed parentheses detected ==> not expand token.
+                    return false;
+                }
+            }
+
+            return count == 0;
+        }
+
+        /// <summary>
+        /// Checks the specified model and structuredType and see whether the token can be resolved to a bound operation name.
+        /// </summary>
+        /// <param name="token">The token to check.</param>
+        /// <returns>true if token can be matched a bound operation; false otherwise.</returns>
+        private bool IsBoundOperationToken(string token)
+        {
+            const char nameSpaceSeparator = '.';
+
+            if (this.edmModel == null)
+            {
+                return false;
+            }
+
+            bool found = false;
+            int idx = token.LastIndexOf(nameSpaceSeparator);
+            if (idx == -1)
+            {
+                // Token is in unqualified form, try matching the unqualified name of the bound operation.
+                found = this.edmModel.FindDeclaredBoundOperations(this.structuredType)
+                        .Any(op => op.Name.Equals(token, StringComparison.Ordinal));
+            }
+            else
+            {
+                // Qualified name: try to match bound and unbound operations
+                found = this.edmModel.FindBoundOperations(token, this.structuredType).Count() != 0
+                    || this.edmModel.FindDeclaredOperations(token).Count() != 0;

```

Comment:
>|| this.edmModel.FindDeclaredOperations(token).Count() != 0 [](start = 20, length = 59)

Do you have a test case for this? I don't think there should ever be unbound operations nested within parens.

Reply:
Yes. In SelectedPropertiesNodeTest, we have tests exercising unbound operations using qualified name such as TestModel.Action(Edm.Int32).

---
In reply to: [224137533](https://github.com/OData/odata.net/pull/1286#discussion_r224137533) [](ancestors = 224137533)

Reply:
not sure that test is valid; the only time an operation should exist within a select list would be if it was bound to the structured type against which the select list is applied.  Otherwise, the fact that there is an unbound operation with the same name is purely coincidental, and we should only resolve it as an operation if it can't be a navigation property (i.e., there is a dot in the name).

---
In reply to: [224629515](https://github.com/OData/odata.net/pull/1286#discussion_r224629515) [](ancestors = 224629515,224137533)

Reply:
The IsBoundOperationToken method has been replaced by the IsNavigationPropertyToken.

---
In reply to: [226138497](https://github.com/OData/odata.net/pull/1286#discussion_r226138497) [](ancestors = 226138497,224629515,224137533)